cranelift/codegen/src/machinst/lower.rs

```diff
@@ -295,15 +295,26 @@ pub struct Lower<'func, I: VCodeInst> {
 /// actually merged" point. Instead, we compute a
 /// transitive-uniqueness. That is what this enum represents.
 ///
+/// There is one final caveat as well to the result of this analysis. Notably
+/// some instructions are considered "root" instructions for this analysis whose
+/// operands don't accurately reflect the use of the results. Root instructions
+/// are currently defined as multi-result instructions. This means that even if
```

Comment:
I would maybe rephrase as:

"Notably, we define some instructions to be "root" instructions, which means that we assume they will always be codegen'd at the root of a matching tree, and not matched. (This comes with the caveat that we actually enforce this property by making them "opaque" to subtree matching in `get_value_as_source_or_const`.) Because they will always be codegen'd once, they in some sense "reset" multiplicity: these root instructions can be used many times, but because their result(s) are only computed once, they only use theirinputs once.

We currently define all multi-result instructions to be "root" instructions, because it is too complex to reason about matching through them, and they cause too-coarse-grained approximation of multiplicity otherwise: the analysis would have to assume (as it used to!) that they are *always* multiply-used, simply because they have multiple outputs even if those outputs are used only once.

In the future we could define other instructions to be "root" instructions as well, if we make the corresponding change to `get_value_as_source_or_const` as well."